### PR TITLE
Configurable restart command

### PR DESCRIPTION
**Pull request by initard**
## Proposed changes

This PR makes the restart command configurable. 

This is the current status of the PR:

- [x] Add key to system.toml
- [x] Move reusable code into library code (from `tedge` to `tedge_config`)
- [x] Implement restart from system.toml in `tedge_agent`
- [x] Fix/add tests   
- [x]  Check that it works. 

For some reason the restart feature does not work on my laptop. Running `sudo -u tedge init 6` fails to restart, stating that the device can not be restarted since there are users logged in. The suggestion is to use `systemctl reboot -i`.  I am not sure if this is only for my laptop or not. I tried also making a debian package for tedge agent and the error there is slightly different: 
```text
Nov 10 14:25:28 pop-os tedge_agent[34932]: sudo: a terminal is required to read the password; either use the -S option to read from standard input or configure an askpass helper
Nov 10 14:25:28 pop-os tedge_agent[34932]: sudo: a password is required
Nov 10 14:25:28 pop-os sudo[34932]: pam_unix(sudo:auth): conversation failed
Nov 10 14:25:28 pop-os sudo[34932]: pam_unix(sudo:auth): auth could not identify password for [tedge]
Nov 10 14:25:28 pop-os sudo[34932]:    tedge : command not allowed ; PWD=/ ; USER=root ; COMMAND=init 6
Nov 10 14:25:28 pop-os tedge_agent[34850]: 2022-11-10T13:25:28.631403597Z ERROR sm-agent: tedge_agent::agent: Command returned non 0 exit code.
``` 

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#1017 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by initard** [comment]
This PR now works with a known bug: https://github.com/thin-edge/thin-edge.io/issues/1135 and https://github.com/thin-edge/thin-edge.io/issues/1579

File: crates/common/tedge_config/src/system_services/managers/config.rs [hunk 1]
```diff
@@ -4,10 +4,12 @@ use std::fs;
 use std::path::PathBuf;
 
 pub const SERVICE_CONFIG_FILE: &str = "system.toml";
+const INIT_COMMAND: &[&str] = &["init", "6"];
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/system_services/managers/config.rs:7]: This name is confusing as there is also an unrelated `InitConfig` struct.

```suggestion
const REBOOT_COMMAND: &[&str] = &["init", "6"];
```

File: crates/common/tedge_config/src/system_services/managers/config.rs [hunk 2]
```diff
@@ -22,6 +24,23 @@ pub struct InitConfig {
     pub is_active: Vec<String>,
 }
 
+#[derive(Deserialize, Debug, Eq, PartialEq)]
+#[serde(deny_unknown_fields)]
+pub struct SystemVariablesConfig {
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/system_services/managers/config.rs:29]: This name provides little help guessing the purpose of this struct.

```suggestion
pub struct SystemSpecificCommands {
```

File: crates/common/tedge_config/src/system_services/managers/config.rs [hunk 3]
```diff
@@ -22,6 +24,23 @@ pub struct InitConfig {
...
+    pub restart: Vec<String>,
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/system_services/managers/config.rs:30]: I would prefer to use the term `reboot` (the device), since the verb `restart` (a service) is already used by `InitConfig`.

File: crates/common/tedge_config/src/system_services/managers/config.rs [hunk 4]
```diff
@@ -4,10 +4,12 @@ use std::fs;
 use std::path::PathBuf;
 
 pub const SERVICE_CONFIG_FILE: &str = "system.toml";
+const INIT_COMMAND: &[&str] = &["init", "6"];
 
 #[derive(Deserialize, Debug, Default, Eq, PartialEq)]
 pub struct SystemConfig {
-    pub(crate) init: InitConfig,
+    pub(crate) init: Option<InitConfig>,
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/system_services/managers/config.rs:11]: Why this change? Why not simply use `InitConfig::default()` when no config is provided?

File: crates/common/tedge_config/src/system_services/managers/general_manager.rs [hunk 5]
```diff
@@ -14,7 +14,8 @@ pub struct GeneralServiceManager {
 
 impl GeneralServiceManager {
     pub fn try_new(config_root: PathBuf) -> Result<Self, SystemServiceError> {
-        let init_config = SystemConfig::try_new(config_root.clone())?.init;
+        let init_config = SystemConfig::try_new(config_root.clone())?.init.unwrap(); // TODO throw
+                                                                                     // error
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/system_services/managers/general_manager.rs:18]: I would simply revert the change making optional the field `SystemConfig::init`.

File: crates/core/tedge_agent/src/agent.rs [hunk 6]
```diff
@@ -661,28 +667,20 @@ impl SmAgent {
...
-
-#[cfg(not(test))]
-fn get_restart_operation_commands() -> Vec<Command> {
+fn get_restart_operation_commands(system_config_path: &Path) -> Result<Vec<Command>, AgentError> {
     let mut vec = vec![];
     // sync first
-    let mut sync_command = std::process::Command::new("sudo");
-    sync_command.arg("sync");
+    let mut sync_command = std::process::Command::new(SUDO);
+    sync_command.arg(SYNC);
```
Comment (didier-wenzek) [inline crates/core/tedge_agent/src/agent.rs:674]: I wonder if this `sync` command can be part of the user configuration.

**Review by didier-wenzek — COMMENTED** [review]
Nice to have reduce to one the number of functions with test/no-test versions.

**Review by didier-wenzek — COMMENTED** [review]
Nice. However, the tests need to be fixed (you failed to revert there the change making `config.init` optional).

File: crates/common/tedge_config/src/system_services/managers/general_manager.rs [hunk 7]
```diff
@@ -15,6 +15,7 @@ pub struct GeneralServiceManager {
 impl GeneralServiceManager {
     pub fn try_new(config_root: PathBuf) -> Result<Self, SystemServiceError> {
         let init_config = SystemConfig::try_new(config_root.clone())?.init;
+
```
Comment (rina23q) [inline crates/common/tedge_config/src/system_services/managers/general_manager.rs:18]: Useless change :D But it's ok

File: crates/common/tedge_config/src/system_services/managers/config.rs [hunk 8]
```diff
@@ -77,6 +96,9 @@ mod tests {
             enable =  ["/bin/systemctl", "enable", "{}"]
             disable =  ["/bin/systemctl", "disable", "{}"]
             is_active = ["/bin/systemctl", "is-active", "{}"]
+
+            [system]
+            restart = "init 6"
```
Comment (rina23q) [inline crates/common/tedge_config/src/system_services/managers/config.rs:101]: `reboot` or `restart`?

**Comment by didier-wenzek** [comment]
> Quickly reviewed (going to do more later). I am not convinced to move `system_services` directory to `/crates/common/tedge_config`. It is definitely not tedge_config, which is related to our `tedge config ...` command. My suggestion is to move directly `/crates/common`.

I agree that moving code related to system config into `tedge_config` change the original purpose of the former. However, we have already too many small crates that makes no sense in a context that is not thin-edge. This is why I prefer to have a larger crate that groups many thin-edge configuration related features.

**Comment by rina23q** [comment]
> I agree that moving code related to system config into tedge_config change the original purpose of the former. However, we have already too many small crates that makes no sense in a context that is not thin-edge. This is why I prefer to have a larger crate that groups many thin-edge configuration related features.

Having one larger crate makes sense. However, how the `tedge_config` crate looks is quite messy because of re-purposing the crate. 

```
├── tedge_config
│   ├── Cargo.toml
│   ├── src
│   │   ├── config_setting.rs
│   │   ├── error.rs
│   │   ├── lib.rs
│   │   ├── models
│   │   │   ├── connect_url.rs
│   │   │   ├── file_path.rs
│   │   │   ├── flag.rs
│   │   │   ├── ipaddress.rs
│   │   │   ├── mod.rs
│   │   │   ├── port.rs
│   │   │   └── templates_set.rs
│   │   ├── settings.rs
│   │   ├── system_services
│   │   │   ├── command_builder.rs
│   │   │   ├── error.rs
│   │   │   ├── manager_ext.rs
│   │   │   ├── manager.rs
│   │   │   ├── managers
│   │   │   │   ├── config.rs
│   │   │   │   ├── general_manager.rs
│   │   │   │   └── mod.rs
│   │   │   ├── mod.rs
│   │   │   └── services.rs
│   │   ├── tedge_config_defaults.rs
│   │   ├── tedge_config_dto.rs
│   │   ├── tedge_config_location.rs
│   │   ├── tedge_config_repository.rs
│   │   └── tedge_config.rs
│   └── tests
│       └── test_tedge_config.rs

```

File: crates/common/tedge_config/src/system_services/managers/config.rs [hunk 9]
```diff
@@ -77,6 +96,9 @@ mod tests {
             enable =  ["/bin/systemctl", "enable", "{}"]
             disable =  ["/bin/systemctl", "disable", "{}"]
             is_active = ["/bin/systemctl", "is-active", "{}"]
+
+            [system]
+            restart = ["init", "6"]
```
Comment (rina23q) [inline crates/common/tedge_config/src/system_services/managers/config.rs:101]: Still, `restart` or `reboot`?

File: crates/common/tedge_config/src/system_services/managers/config.rs [hunk 10]
```diff
@@ -107,6 +133,9 @@ mod tests {
             enable =  ["/bin/systemctl", "enable", "{}"]
             disable =  ["/bin/systemctl", "disable", "{}"]
             is_active = ["/bin/systemctl", "is-active", "{}"]
+
+            [system]
+            restart = "init 6"
```
Comment (rina23q) [inline crates/common/tedge_config/src/system_services/managers/config.rs:138]: Another location, `restart` or `reboot`?

**Comment by didier-wenzek** [comment]
> > I agree that moving code related to system config into tedge_config change the original purpose of the former. However, we have already too many small crates that makes no sense in a context that is not thin-edge. This is why I prefer to have a larger crate that groups many thin-edge configuration related features.
> 
> Having one larger crate makes sense. However, how the `tedge_config` crate looks is quite messy because of re-purposing the crate. In my opinion, if we make a new directory `tedge_config` and put all `tedge_config_*.rs` files to there, it will look better.

It's a good idea to add a `tedge_config` sub-module. However, I would do that in another different PR.

**Review by didier-wenzek — APPROVED** [review]
Approved
